src/io/binary_parser.cpp

```diff
@@ -119,6 +109,13 @@ long long read_variables(string filename, dd::FactorGraph &fg) {
     fg.variables[id] =
         dd::Variable(id, type_const, is_evidence, 0, upper_bound, init_value,
                      init_value, edge_count, is_observation);
+
+    // set up the default domains for multinomial
+    for (int i = 0; i < cardinality; i++) {
+      fg.variables[id].domain.push_back(i);
+      fg.variables[id].domain_map[i] = i;
+    }
```

Comment:
I'm afraid creating a vector and a map for every variable here will be disastrous.  In the dense case, copies of identical vector and map will be allocated for all of them.  And in the sparse case, they'll immediately get cleared when the domains are loaded.  Can we replace the parts that rely on these maps and vectors with perhaps inline functions that returns `i` when asked for `i` if no domain was loaded?